src/bin/pgcopydb/ld_stream.c

```diff
@@ -35,6 +35,150 @@
 static bool updateStreamCounters(StreamContext *context,
 								 LogicalMessageMetadata *metadata);
 
+/*
+ * Determines whether a message needs to be filtered out based on the
+ * provided filters. The caller function should have already parsed and
+ * validated the filters.
+ */
+bool
+ShouldFilterOutMessage(SourceFilters *filters, char *nspname, char *relname)
+{
+	/*
+	 * Validate nspname is not NULL or empty
+	 */
+	if (nspname == NULL || IS_EMPTY_STRING_BUFFER(nspname))
+	{
+		log_error("BUG: nspname is NULL or empty");
+		return false;
+	}
+
+	/*
+	 * Validate relname is not NULL or empty
+	 */
+	if (relname == NULL || IS_EMPTY_STRING_BUFFER(relname))
+	{
+		log_error("BUG: relname is NULL or empty");
+		return false;
+	}
+
+	/*
+	 * Filter-out pgcopydb.* in any case
+	 */
+	if (streq(nspname, "pgcopydb"))
+	{
+		log_info("Filtering out message for pgcopydb internal tables \"%s\"", nspname);
+		return true;
+	}
+
+	/*
+	 * If no filters are set, then we don't need to filter out any messages
+	 */
+	if (filters->type == SOURCE_FILTER_TYPE_NONE)
+	{
+		return false;
+	}
+
+	/*
+	 * Filtering based on the sourceFilters.exclude-table-data
+	 * return true if the table is in the list
+	 * exclude-table-data filters should be processed first, otherwise
+	 * include-only-table will be able to filter the message before
+	 * we can check if the table is in the exclude-table-data list
+	 * and we will not be able to filter out the message.
+	 */
```

Comment:
Please review the filtersDB SQLite catalog, I believe you could implement a single look-up in the main filter table, or maybe in the s_table table and have all the information you need in a single lookup.